File: src/java/test/src/androidTest/java/com/tencent/wcdbtest/base/TableTestCase.java
// Created by qiuwenchen on 2023/5/8.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.tencent.wcdbtest.base;

import com.tencent.wcdb.base.Value;
import com.tencent.wcdb.base.WCDBException;
import com.tencent.wcdb.core.Table;
import com.tencent.wcdb.orm.TableBinding;
import com.tencent.wcdb.winq.Column;
import com.tencent.wcdb.winq.ColumnDef;
import com.tencent.wcdb.winq.ColumnType;
import com.tencent.wcdb.winq.StatementCreateTable;

import org.junit.Assert;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class TableTestCase extends DatabaseTestCase {
    public String tableName = "testTable";
    public TableBinding<TestObject> tableBinding = DBTestObject.INSTANCE;
    public boolean isVirtualTable = false;
    public Table<TestObject> table;

    public void createTable() throws WCDBException {
        if (!isVirtualTable) {
            database.createTable(tableName, tableBinding);
        } else {
            database.createVirtualTable(tableName, tableBinding);
        }
        table = database.getTable(tableName, tableBinding);
    }

    public void createValueTable() {
        StatementCreateTable createTable = new StatementCreateTable().createTable(tableName);
        createTable.define(new ColumnDef("id", ColumnType.Integer).makePrimary(true));
        createTable.define(new ColumnDef("content", ColumnType.Text));
        database.execute(createTable);
    }

    public Column[] columns() {
        return new Column[]{new Column("id"), new Column("content")};
    }

    public void createVirtualTable() throws WCDBException {
        database.createVirtualTable(tableName, tableBinding);
        table = database.getTable(tableName, tableBinding);
    }

    public void dropTable() throws WCDBException {
        database.dropTable(tableName);
    }

    public void doTestObjectsAfterInsert(Object[] objects, int insertCount, String[] sqls, TestOperation operation) {
        if(insertCount > 1) {
            ArrayList<String> list = new ArrayList<>(Arrays.asList(sqls));
            list.add(0, "BEGIN IMMEDIATE");
            list.add("COMMIT");
            sqls = list.toArray(new String[list.size()]);
        }
        doTestObjectsAfterOperation(objects, sqls, operation);
    }

    public void doTestObjectAfterOperation(Object object, String sql, TestOperation operation) {
        doTestObjectsAfterOperation(new Object[]{object}, new String[]{sql}, operation);
    }

    public void doTestObjectsAfterOperation(Object[] objects, String sql, TestOperation operation) {
        doTestObjectsAfterOperation(objects, new String[]{sql}, operation);
    }

    public void doTestObjectsAfterOperation(Object[] objects, String[] sqls, TestOperation operation) {
        doTestSQLs(sqls, operation);
        ArrayList<TestObject> allObjects = getAllObjects();
        Assert.assertTrue(((objects == null || objects.length == 0) &&
                (allObjects == null || allObjects.size() == 0)) ||
                Arrays.equals(objects, allObjects.toArray()));
    }

    public interface SelectingObjectOperation {
        List execute();
    }

    public void doTestObjectBySelecting(Object object, String sql, SelectingObjectOperation operation) {
        doTestObjectBySelecting(new Object[]{object}, new String[]{sql}, operation);
    }

    public void doTestObjectBySelecting(Object[] objects, String sql, SelectingObjectOperation operation) {
        doTestObjectBySelecting(objects, new String[]{sql}, operation);
    }

    public void doTestObjectBySelecting(Object[] objects, String[] sqls, SelectingObjectOperation operation) {
        ArrayList selecting = new ArrayList();
        doTestSQLs(sqls, new TestOperation() {
            @Override
            public void execute() throws WCDBException {
                selecting.addAll(operation.execute());
            }
        });
        Assert.assertTrue(((objects == null || objects.length == 0) &&
                (selecting == null || selecting.size() == 0)) ||
                Arrays.equals(objects, selecting.toArray()));
    }

    public interface SelectingRowsOperation {
        ArrayList<Value[]> execute();
    }

    public void doTestValueBySelecting(Value value, String sql, SelectingRowsOperation operation) {
        doTestRowsBySelecting(Collections.singletonList(new Value[]{value}), sql, operation);
    }

    public void doTestColumnBySelecting(List<Value> column, String sql, SelectingRowsOperation operation) {
        ArrayList<Value[]> rows = new ArrayList<>();
        for(Value value : column) {
            rows.add(new Value[]{value});
        }
        doTestRowsBySelecting(rows, new String[]{sql}, operation);
    }

    public void doTestRowBySelecting(Value[] row, String sql, SelectingRowsOperation operation) {
        doTestRowsBySelecting(Collections.singletonList(row), new String[]{sql}, operation);
    }

    public void doTestRowsBySelecting(List<Value[]> rows, String sql, SelectingRowsOperation operation) {
        doTestRowsBySelecting(rows, new String[]{sql}, operation);
    }

    public void doTestRowsBySelecting(List<Value[]> rows, String[] sqls, SelectingRowsOperation operation) {
        ArrayList selecting = new ArrayList();
        doTestSQLs(sqls, new TestOperation() {
            @Override
            public void execute() throws WCDBException {
                selecting.addAll(Arrays.asList(operation.execute()));
            }
        });
        Assert.assertTrue(((rows == null || rows.size() == 0) &&
                (selecting == null || selecting.size() == 0)) ||
                Arrays.equals(rows.toArray(), selecting.toArray()));
    }

    public ArrayList<TestObject> getAllObjects() throws WCDBException {
        return table.getAllObjects();
    }
}

File: src/kotlin/test/src/androidTest/java/com/tencent/wcdbtest/base/TableTestCase.kt
// Created by chenqiuwen on 2023/6/24.
//

/*
 * Tencent is pleased to support the open source community by making
 * WCDB available.
 *
 * Copyright (C) 2017 THL A29 Limited, a Tencent company.
 * All rights reserved.
 *
 * Licensed under the BSD 3-Clause License (the "License"); you may not use
 * this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 *       https://opensource.org/licenses/BSD-3-Clause
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.tencent.wcdbtest.base

import com.tencent.wcdb.base.WCDBException
import com.tencent.wcdb.core.Table
import com.tencent.wcdb.orm.TableBinding
import org.junit.Assert
import java.util.*

typealias SelectingObjectOperation = () -> List<TestObject>

open class TableTestCase : DatabaseTestCase() {
    var tableName = "testTable"
    var tableBinding: TableBinding<TestObject> = DBTestObject
    var isVirtualTable = false
    lateinit var table: Table<TestObject>

    @Throws(WCDBException::class)
    fun createTable() {
        if (!isVirtualTable) {
            database.createTable(tableName, tableBinding)
        } else {
            database.createVirtualTable(tableName, tableBinding)
        }
        table = database.getTable(tableName, tableBinding)
    }

    @Throws(WCDBException::class)
    fun createVirtualTable() {
        database.createVirtualTable(tableName, tableBinding)
        table = database.getTable(tableName, tableBinding)
    }

    @Throws(WCDBException::class)
    fun dropTable() {
        database.dropTable(tableName)
    }

    fun doTestObjectsAfterInsert(
        objects: Array<Any>,
        insertCount: Int,
        sqls: Array<String>,
        operation: TestOperation
    ) {
        var newSqls = sqls
        if (insertCount > 1) {
            val list: MutableList<String> = mutableListOf(*sqls)
            list.add(0, "BEGIN IMMEDIATE")
            list.add("COMMIT")
            newSqls = list.toTypedArray()
        }
        doTestObjectsAfterOperation(objects, newSqls, operation)
    }

    fun doTestObjectsAfterOperation(
        objects: Array<Any>,
        sql: String,
        operation: TestOperation
    ) {
        doTestObjectsAfterOperation(objects, arrayOf(sql), operation)
    }

    fun doTestObjectsAfterOperation(
        objects: Array<Any>,
        sqls: Array<String>,
        operation: TestOperation
    ) {
        doTestSQLs(sqls, operation)
        val allObjects: List<TestObject> = try {
            this.allObjects
        } catch (e: WCDBException) {
            throw RuntimeException(e)
        }
        Assert.assertTrue(
            (objects.isEmpty() && allObjects.isEmpty()) ||
                    objects.contentEquals(allObjects.toTypedArray())
        )
    }

    fun doTestObjectBySelecting(
        `object`: TestObject,
        sql: String,
        operation: SelectingObjectOperation
    ) {
        doTestObjectBySelecting(arrayOf(`object`), arrayOf(sql), operation)
    }

    fun doTestObjectBySelecting(
        objects: Array<TestObject>,
        sql: String,
        operation: SelectingObjectOperation
    ) {
        doTestObjectBySelecting(objects, arrayOf(sql), operation)
    }

    fun doTestObjectBySelecting(
        objects: Array<TestObject>,
        sqls: Array<String>,
        operation: SelectingObjectOperation
    ) {
        val selecting = mutableListOf<TestObject>()
        doTestSQLs(sqls) {
            selecting.addAll(operation())
        }
        Assert.assertTrue(
            (objects.isEmpty() && selecting.isEmpty()) ||
                    objects.contentEquals(selecting.toTypedArray())
        )
    }

    @get:Throws(WCDBException::class)
    val allObjects: List<TestObject>
        get() = table.allObjects
}